scipy/interpolate/regrid_python.py

```diff
@@ -0,0 +1,1174 @@
+        Ay, offs_y, Dy, offs_dy, ncy, ky, p)
+
+    if p != -1:
+        Q = np.vstack([Q, np.zeros((Dx.shape[0], Q.shape[1]), dtype=np.float64)])
+
+    _dierckx.qr_reduce(Ax_aug, offset_aug_x, nc_augx, Q)
+
+    Q_t = np.ascontiguousarray(Q)
+    cT, _, _ = _dierckx.fpback(
+        Ax_aug, nc_augx, x_x,
+        Q_t, tx, kx, w_x,
+        Q_t, False
+    )
+    W = cT.T
+
+    if p != -1:
+        W = np.vstack([W, np.zeros((Dy.shape[0], W.shape[1]), dtype=np.float64)])
+
+    WT = np.ascontiguousarray(W)
+    _dierckx.qr_reduce(Ay_aug, offset_aug_y, nc_augy, WT)
+    W1 = np.ascontiguousarray(WT)
+
+    C, _, fp = _dierckx.fpback(
+        Ay_aug, nc_augy,
+        x_y, W1, ty, ky, w_y,    # y = W1 (dummy)
+        W1,                      # yw = RHS = W1 → returns C
+        False
+    )
+
+    C = C.T
+    _Ax = BSpline.design_matrix(x_x, tx, kx, extrapolate=False)
+    _Ay = BSpline.design_matrix(x_y, ty, ky, extrapolate=False)
+    zhat = (_Ax @ C) @ _Ay.T
+    fp = fp_residual(z, zhat)
+    return C, fp
+
+def _span_indices_from_knots(x, t, k):
+    """
+    Map each ``x`` to its B-spline span index (clamped).
+
+    Parameters
+    ----------
+    x : array_like
+        Query points.
+    t : array_like
+        Knot vector.
+    k : int
+        Spline degree.
+
+    Returns
+    -------
+    ndarray of int
+        Span indices ``j`` such that ``t[j] <= x < t[j+1]``, clamped to
+        ``[k, len(t) - k - 2]``.
+    """
+    j = np.searchsorted(t, x, side="right") - 1
+    j = np.clip(j, k, len(t) - k - 2)
+    return j
+
+def _fpint_axis_from_residual_rows(x, t, k, row_energy):
+    """
+    Accumulate residual row energies into knot-span bins.
+
+    Parameters
+    ----------
+    x : array_like
+        Axis sample positions.
+    t : array_like
+        Knot vector along the same axis.
+    k : int
+        Spline degree.
+    row_energy : array_like, shape (len(x),)
+        Energy per row, e.g. ``sum(R**2, axis=1)``.
+
+    Returns
+    -------
+    ndarray
+        ``fpint`` array of length ``len(t) - 1`` with per-span energy totals.
+        Zero for degenerate spans where ``t[i+1] == t[i]``.
+    """
+    t = np.asarray(t, float)
+    valid_span = (t[1:] > t[:-1])
+    fpintx = np.zeros(len(t) - 1, dtype=float)
+
+    spans = _span_indices_from_knots(x, t, k)
+    np.add.at(fpintx, spans, row_energy)
+
+    fpintx[~valid_span] = 0.0
+    return fpintx
+
+def _fpint_xy(x, y, t_x, t_y, kx, ky, R):
+    """
+    Project residual energy along x and y knot spans.
+
+    Parameters
+    ----------
+    x, y : array_like
+        Sample positions along x and y.
+    t_x, t_y : array_like
+        Knot vectors along x and y.
+    kx, ky : int
+        Spline degrees.
+    R : ndarray, shape (len(x), len(y))
+        Residual grid ``Z - Zhat``.
+
+    Returns
+    -------
+    fpintx, fpinty : ndarray
+        Per-span energy arrays for x and y axes, respectively.
+    """
+    row_energy = np.einsum("ij,ij->i", R, R)
```

Comment:
Note to self: Is this `trace(R.mT @ R)`? From reading the code, I can't tell what this function is doing.